containers/src/Kokkos_UnorderedMap.hpp

```diff
@@ -174,10 +204,20 @@ class UnorderedMapInsertResult {
 /// \tparam EqualTo Definition of the equality function for instances of
 ///   <tt>Key</tt>.  The default will do a bitwise equality comparison.
 ///
+/// \tparam InsertOp Definition of the insert function for duplicate
+///   insertions of <tt>Key</tt>. The default is to set value to the
+///   first key, value pair that is inserted. Please see
+///   Kokkos::UnorderedMapInsertOpTypes for more details.
+///
 template <typename Key, typename Value,
-          typename Device  = Kokkos::DefaultExecutionSpace,
-          typename Hasher  = pod_hash<std::remove_const_t<Key>>,
-          typename EqualTo = pod_equal_to<std::remove_const_t<Key>>>
+          typename Device   = Kokkos::DefaultExecutionSpace,
+          typename Hasher   = pod_hash<std::remove_const_t<Key>>,
+          typename EqualTo  = pod_equal_to<std::remove_const_t<Key>>,
+          typename InsertOp = typename UnorderedMapInsertOpTypes<
+              Kokkos::View<std::remove_const_t<std::conditional_t<
+                               std::is_void_v<Value>, int, Value>> *,
+                           Device>,
+              uint32_t>::NoOp>
```

Comment:
Adding a template parameter is a breaking change we can't do that really. We could discuss adding this only if deprecated code off, but we can't simply add it. Alternatively, why can't we have an overload of insert which takes an insert op? Is there a reason this has to be a global template for the class?

Reply:
This seems like the lightest-weight way to add this operation as well as following the implementation style of UnorderedMap. What about keeping the current approach with a single constructor overload, rather than a new insert method, the constructor overload could include the InsertOp type and constructor parameter?

Reply:
Thinking about this more now, I see no reason for making the InsertOp a global template for the class other than following the existing implementation style of UnorderedMap (e.g. EqualTo, etc.). I will try refactoring with a new insert method instead. Can you elaborate on how the InsertOp template parameter is a breaking change?

Reply:
After trying to remove the InsertOp as a global template parameter, I found that deep copy no longer works. I will look into using insert method overloads rather than a class member.